### PR TITLE
TS-29451 Collector Control API

### PR DESCRIPTION
**Pull request by stahlbauer**
Addresses issue [TS-29451](https://cqse.atlassian.net/browse/TS-29451)

- [ ] Changes are tested adequately
- [ ] Profilers's README.md and the Teamscale documentation are updated in case of user-visible changes
- [ ] CHANGELOG.md of the affected packages are updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

### CODE REVIEW

File: packages/teamscale-coverage-collector/test/system/ControlServer.test.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
+import { App } from '../../src/App';
```
Comment (andipunz) [inline packages/teamscale-coverage-collector/test/system/ControlServer.test.ts:1]: Maybe add a dedicated test for each endpoint? Switch partition, project, ... ?

File: packages/teamscale-coverage-collector/test/CollectorClient.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import WebSocket from 'ws';
```
Comment (andipunz) [inline packages/teamscale-coverage-collector/test/CollectorClient.ts:1]: Is this a test class? If yes stick to the CollectorClient.test.ts naming as the other tests are named?

Comment (karottenreibe) [reply packages/teamscale-coverage-collector/test/CollectorClient.ts:1]: no, this is not a test but rather utilities to be used in tests, so I think the file naming is adequate

File: packages/teamscale-coverage-collector/test/system/ControlServer.test.ts (hunk 1 above)
Comment (karottenreibe) [reply packages/teamscale-coverage-collector/test/system/ControlServer.test.ts:1]: Thanks for that, I actually found 1 bug and several inconsistencies this way :-)

**Comment by karottenreibe** [comment]
rfr again

File: packages/teamscale-coverage-collector/src/App.ts [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+ * Error that is thrown when the upload to Teamscale failed
+ */
+class TeamscaleUploadError extends Error {
+	// No special fields needed compared to Error.
+	// Sole use is to be able to distinguish Teamscale upload errors from other errors.
+}
+
+/**
+ * The main class of the Teamscale JavaScript Collector.
+ * Used to start the collector for with a given configuration.
```
Comment (andipunz) [inline packages/teamscale-coverage-collector/src/App.ts:28]:
```suggestion
 * Used to start the collector with a given configuration.
```

File: packages/teamscale-coverage-collector/src/App.ts [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+		// ATTENTION: The server is executed asynchronously
+		const serverState = server.start();
+
+		// Optionally, start a timer that dumps the coverage after a N seconds
+		const timerState = this.maybeStartDumpTimer(config, storage, logger);
+
+		// Say bye bye on CTRL+C and exit the process
+		process.on('SIGINT', async () => {
+			// ... and do a final dump before.
+			await this.dumpCoverage(config, storage, logger).then();
```
Comment (andipunz) [inline packages/teamscale-coverage-collector/src/App.ts:103]: is this .then() really needed with the await keyword?

Comment (andipunz) [reply packages/teamscale-coverage-collector/src/App.ts:103]:
```suggestion
			await this.dumpCoverage(config, storage, logger);
```

File: packages/teamscale-coverage-collector/src/App.ts [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+	 */
+	private static maybeStartDumpTimer(
+		config: ConfigParameters,
+		storage: DataStorage,
+		logger: Logger
+	): { stop: () => void } {
+		if (config.dump_after_mins > 0) {
+			logger.info(`Will dump coverage information every ${config.dump_after_mins} minute(s).`);
+			const timer = setInterval(() => {
+				this.dumpCoverage(config, storage, logger).then();
```
Comment (andipunz) [inline packages/teamscale-coverage-collector/src/App.ts:134]: consistently use async/await?

File: packages/teamscale-coverage-collector/src/App.ts [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+					if (response.status >= 400) {
+						throw new TeamscaleUploadError(
+							`Upload failed with code ${response.status}: ${response.statusText}. Response Data: ${response.data}`
+						);
+					} else {
+						logger.info(`Upload with status code ${response.status} finished.`);
+					}
+				} else if (error.request) {
+					throw new TeamscaleUploadError(`Upload request did not receive a response.`);
+				}
```
Comment (andipunz) [inline packages/teamscale-coverage-collector/src/App.ts:242]: Here the async/await is also mixed up with the .then() and .catch()
I think when using await it should be something like:
```
try{
let response = await axios.post(...);
logger.info(`Upload finished with code ${response.status}.`);
}catch(...){
...
}
```

**Review by andipunz — APPROVED** [review]

**Pull request merged**